### PR TITLE
Support versioned event appliers

### LINKED ISSUE #12764: Version event appliers
**Description**

We need to introduce a way to version our Event Appliers. 

This is required for:
- #11456 

Currently, the `Deployment.CREATED` event applier inserts a deployment record into the state. This is used to be able to re-distribute the deployment later in case distribution failed. This deployment record is deleted (cleaned up) by the `Deployment.FULLY_DISTRIBUTED` event applier.

With generalized record distribution, we will still write the `Deployment.CREATED` events, but there is no need anymore for this deployment record because it is stored for distribution in a generalized way (see #11456). In addition, we no longer write the `Deployment.FULLY_DISTRIBUTED` event. So, we cannot clean up the state anymore if we would continue to insert a deployment record on `Deployment.CREATED`.

After updating to a version with generalized record distribution, unprocessed deployment distribution commands may exist on the log, and deployment distribution events must be replayed in the same way as they were when they were written. So, in order to stay backward compatible, both the existing Deployment Distribution mechanism and the Generalized Record Distribution need to co-exist. 

To allow the co-existence of these mechanisms, we need to have a way to determine how the event should be applied. This can be done by versioning our event appliers.

**Considered alternatives**

1. **Continue to write the `Deployment.FULLY_DISTRIBUTED` event for the Generalized Record Distribution of deployments**. This would allow us to keep the `Deployment.CREATED` event applier as is, continuing to insert the deployment record into the state. This alternative has some downsides:
- we are inserting the deployment record twice into the state (once redundantly), which is a performance regression (it was only once before).
- the `Deployment.FULLY_DISTRIBUTED` event leaks the distribution responsibility into deployments, which is what we were trying to remove with generalized record distribution.
- there is no possibility to remove the `Deployment.FULLY_DISTRIBUTED` event in future releases, exactly for the problem above. We'd need to change the `Deployment.CREATED` event applier, which is not possible without versioning.

2. **Replace `Deployment.CREATED` with a `Deployment.CREATED_V2` intent**. Intents are the standard way to change the behavior of processing/applying records. This alternative has some downsides:
- it's not pretty to have a version number in the intent. We could avoid this by carefully thinking about alternative intent names, like `DEPLOYED` but it would never fit the `CREATE` command.
- users might depend on consuming the `Deployment.CREATED` event in an exporter to extract the deployed resources.

3. **Stop writing the `Deployment.CREATED` event altogether**. The event applier can stay as is, to replay previously written events. This alternative has some downsides:
- each command must be followed up by a follow-up record to indicate that it was processed. A `Deployment.CREATE` command without resources is rejected. Those with acceptable resources will always produce `Process.CREATED` or `Decision.CREATED` events. However, it is not so nice if the `Deployment.CREATE` command does not have `Deployment.CREATED` follow-up event.
- users might depend on consuming the `Deployment.CREATED` event in an exporter to extract the deployed resources.

### PR DESCRIPTION
**Pull request by korthout**
## Description

Adds the ability to register event appliers for the combination of intent and record version. Events are applied by the state writer using the record version that is chosen when the event is appended. And events that were appended are applied using their record version when replayed.

## Related issues

closes #12764 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: engine/src/test/java/io/camunda/zeebe/engine/processing/variable/EventApplyingStateWriter.java [hunk 1]
```diff
@@ -34,7 +35,13 @@ public EventApplyingStateWriter(
 
   @Override
   public void appendFollowUpEvent(final long key, final Intent intent, final RecordValue value) {
-    eventWriter.appendFollowUpEvent(key, intent, value);
+    appendFollowUpEvent(key, intent, value, RecordMetadata.DEFAULT_RECORD_VERSION);
```
[the remark this answers is not in the document]
Comment (remcowesterhoud) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/variable/EventApplyingStateWriter.java:38]: As discussed outside of Github, we should do that. I think it also helps prevent mistakes as when creating new features we don't have to think about versioning, as it uses the latest by default.
Downside is, when adding a new applier if we forget to update the version of all the old events, stuff will break. However, I do believe this is more of an edge case than adding new features.

We can do this in a later PR though.

File: engine/src/main/java/io/camunda/zeebe/engine/state/appliers/EventAppliers.java [hunk 2]
```diff
@@ -293,13 +294,22 @@ private void registerCommandDistributionAppliers(final MutableProcessingState st
...
+  public void applyState(
+      final long key, final Intent intent, final RecordValue value, final int recordVersion) {
     final var eventApplier =
-        mapping.getOrDefault(intent, UNIMPLEMENTED_EVENT_APPLIER.apply(intent));
+        mapping.getOrDefault(
+            new IntentAndVersion(intent, recordVersion), UNIMPLEMENTED_EVENT_APPLIER.apply(intent));
     eventApplier.applyState(key, value);
   }
+
+  record IntentAndVersion(Intent intent, int version) {}
```
Comment (remcowesterhoud) [inline engine/src/main/java/io/camunda/zeebe/engine/state/appliers/EventAppliers.java:315]: 💭 Thinking ahead here. If we want to use the latest event applier by default we can just write a compare function for this record and search the map for the latest version of a specific intent.

**Review by remcowesterhoud — APPROVED** [review]
Not much to add except some thoughts 🚀 Let's get the record version merged so we can continue with this straight after

**Comment by remcowesterhoud** [comment]
@korthout Are we good to merge this one? I'd already reviewed it and the changes look good to me.

**Comment by korthout** [comment]
@remcowesterhoud Did you see and read b981024? If so, let's merge this 👍

**Pull request merged**